---
title: "Monte Carlo Sampling from a Beta-Distribution Posterior"
output: html_notebook
---

In the previous session, we encountered the Beta distribution and 
considered how to determine the parameters of a Beta posterior distribution
for allele frequencies.  

For review, if $q$, the frequency of the `1` allele in a poplution has a Beta prior
with parameters $a$ and $b$, we say:
$$
q \sim \mathrm{Beta}(a, b)
$$
And, if $x$ counts the number of `1` alleles at a marker found in our sample of size $N$ gene copies from the population (for example, of elephants). Then the posterior probability
distribution for $q$ is also a Beta distribution:
$$
p(q|x) \propto p(q|x)p(q) = q^{x + a - 1}(1-q)^{N-x+b - 1}.
$$
Such a distribution can be recognized as a Beta distribution with parameters
$x+a$ and $N - x + b$:
$$
q|x \sim \mathrm{Beta}(x+a, N-x+b).
$$

In the previous session, you wrote a function to compute the posterior
distributions of the allele frequencies in $K$ different populations using 
a sample of individual genotypes, and a vector telling us which population
each of those individuals belonged to.

As we will see, that is not the endpoint of a genetic mixture analysis, but we will
pause here and consider Monte Carlo sampling in the context of having a posterior
distribution for the frequency $q$ of the `1` allele in a single population.

Imagine we are estimating the frequency of the `1` allele when we have sampled 50 _diploid_
individuals from a population.  This means we have a sample of 100 gene copies.
For the purposes of estimating $q$, in a population
that is in Hardy-Weinberg equilibrium, this is similar to
sampling $N=100$ _haploid_ individuals.

Suppose 30 of those gene copies were of allelic type `1` and the remaining 70 were
of allelic type `0`.

Further, assume a uniform prior, (i.e., a $\mathrm{Beta}(1,1)$ prior distribution)
on $q$, the frequency of the `1` allele.

As we saw earlier, such a setup yields a _posterior distribution_ for the
allele frequency that is a $\mathrm{Beta}(31, 71)$ distribution.

Various summaries of this posterior (posterior mean, posterior median,
credible intervals, etc.) are analytically avaiable; however, we will practice
obtaining them (and, further below, more elaborate summaries) using Monte
Carlo sampling.

A sample from a Beta distribution can be obtained from R's `rbeta()` function.
Thus the following gives a Monte Carlo sample of size $n = 1000$ from a
$\mathrm{Beta}(31,71)$ distribution:
```{r}
S <- rbeta(1000, 31, 71)
```

# Starting exercises

1. First get a visual approximation of the posterior distribution (as a histogram)
using a Monte Carlo sample of size 1,000,000.
```{r}
hist(rbeta(1e6, 31, 71), breaks = seq(-0.01, 1.01, by = 0.01))
```
2. Find the posterior mean of the allele frequency from 5 different Monte Carlo samples of size 1000.
```{r}
lapply(1:5, function(x) mean(rbeta(1000, 31, 71)))    
```
3. Do the same for 5 different Monte Carlo samples of size 100
```{r}
lapply(1:5, function(x) mean(rbeta(100, 31, 71)))    
```
4. Find the posterior median.
```{r}
median(rbeta(1000, 31, 71))
```
5. Evaluate the posterior probability that the allele frequency is greater than 0.35.  Do this
for multiple Monte Carlo samples of size 1000 and 100.
```{r}
lapply(1:5, function(x) mean(rbeta(1000, 31, 71) > 0.35))
lapply(1:5, function(x) mean(rbeta(100, 31, 71) > 0.35))
```
6. For each of the above, identify what we called $g(x)$ in the lecture [though, in this context
it is really $g(q)$]---the function of the random variable whose expectation
is being evaluated.

7. The quantiles of a Monte Carlo sample from a posterior distribution
can provide an estimate of credible intervals.
Approximate the equal-tail, 90% credible interval for the allele frequency.
```{r}
quantile(rbeta(10000, 31, 71), probs = c(0.05, 0.95))
```


# Further exercises

We have the posterior distribution of the frequency of the `1` allele, but we might be
more interested in the frequency of individuals with the genotype `11`; i.e., those
diploid individuals that have two copies of the `1` allele.  Assume that the population
is in Hardy-Weinberg equilibrium.
If a population is in Hardy Weinberg equilibrium and allele `1` is at a frequency of $q$, then
the expected frequency of `11` homozygotes is $q^2$.  

8. Use a Monte Carlo sample of size $n = 1,000,000$ of the allele
frequency to make an approximation (as a
histogram) of the posterior distribution of the frequency of `11` homozygotes.
```{r}
S <- rbeta(1e6, 31, 71)
hist(S ^ 2, breaks = seq(-0.01, 1.01, by = 0.01))
```

9. Find the posterior probability that the frequency of `11` homozygotes (assuming H-W equilibrium)
is greater than 0.15.
```{r}
mean(rbeta(1e6, 31, 71) ^ 2 > 0.15)
```


10. Given the posterior for the frequency of the `1` allele, and assuming H-W equilibrium, evaluate
the posterior probability that a new and different sample of 10 individuals from this
population will **not** include
a `11` homozygote.
```{r}
S <- rbeta(1e6, 31, 71)

# for the sample to not include a 11 homozygote, that means that each time we
# sampled an individual it was not a 11.  The probability that an individual is
# 11 is q^2 so the probability that it is not 11 is 1 - q^2.  The probability
# that all 10 are not 11 is thus (1 - q^2)^10. So...
mean( (1 - S ^ 2) ^ 10)
```

These final three exercises illustrate one of the convenient aspects of
Monte Carlo sampling from a posterior.  If you are
interested in the posterior distribution of some _function_ of the variable you have the posterior
for, you can evaluate that directly by sampling the posterior you have in hand.  This is often much
easier than trying to analytically derive the posterior for your function via methods for finding the
distribution of transformed random variables.